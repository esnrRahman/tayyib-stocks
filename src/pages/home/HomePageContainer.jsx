import React, { Component, Fragment } from 'react';
import { Dropdown } from 'semantic-ui-react';

import { StyledHomePageDiv, StyledImageHolder } from './HomePageContainerStyles';
import TayyibStocksLogo from './tstocks_full_logo.png';
import {
  THRESHOLD, getCalculationFromYahoo, getCompanySymbolOptionsFromYahoo,
} from '../../services/yahoo-finance-api-client';
import ResultMessage from './components/ResultMessage';
import ResultChart from './components/ResultChart';

class HomePageContainer extends Component {
  constructor(props) {
    super(props);

    this.state = {
      companyOptions: [],
      chartValues: null,
      resultMessageProps: null,
    };
  }

  handleOnChange = async (event, data) => {
    event.preventDefault();
    this.setState({ resultMessageProps: null });
    const yahooResult = await getCalculationFromYahoo(data.value);
    this.setState({ resultMessageProps: this.getResultMessage(yahooResult) });
  }

  handleOnSearchChange = async (event, data) => {
    event.preventDefault();
    const companyOptions = await getCompanySymbolOptionsFromYahoo(data.searchQuery);
    this.setState({ companyOptions });
  }

  getChartValues = (yahooResult) => {
    const TO_DECIMAL_PLACES = 2;

    const debtValue = parseFloat(yahooResult.totalLiabilitiesPercentage.toFixed(TO_DECIMAL_PLACES));
    const cashValue = parseFloat(
      yahooResult.cashAndShortTermInvestmentsPercentage.toFixed(TO_DECIMAL_PLACES)
    );
    const receivablesValue = parseFloat(
      yahooResult.totalReceivablesPercentage.toFixed(TO_DECIMAL_PLACES)
    );

    const chartValues = {};

    chartValues.debtPass = debtValue < THRESHOLD ? debtValue : THRESHOLD;
    chartValues.debtFail = debtValue < THRESHOLD
      ? 0
      : parseFloat((debtValue - THRESHOLD).toFixed(TO_DECIMAL_PLACES));
    chartValues.cashPass = cashValue < THRESHOLD ? cashValue : THRESHOLD;
    chartValues.cashFail = cashValue < THRESHOLD
      ? 0
      : parseFloat((cashValue - THRESHOLD).toFixed(TO_DECIMAL_PLACES));
    chartValues.receivablesPass = receivablesValue < THRESHOLD ? receivablesValue : THRESHOLD;
    chartValues.receivablesFail = receivablesValue < THRESHOLD
      ? 0
      : parseFloat((cashValue - THRESHOLD).toFixed(TO_DECIMAL_PLACES));

    return chartValues;
  }

  getResultMessage = (result) => {
    let resultMessageProps;

    switch (true) {
      case ('isFound' in result && !result.isFound):
        resultMessageProps = {
          title: 'ERROR',
          description: 'Stock not found',
          isFailure: true,
        };
        break;
      case ('doSplitsExist' in result && result.doSplitsExist):
        resultMessageProps = {
          title: 'FAILURE',
          description: 'Could not calculate financial ratios properly due to technical issues',
          isFailure: true,
        };
        break;
      case ('doAllPass' in result && result.doAllPass): {
        this.setState({ chartValues: this.getChartValues(result) });
        resultMessageProps = {
          title: 'SUCCESS',
          description: 'Stock passes financial ratio screens; please ensure it passes industry/business screens before investing',
          isFailure: false,
        };
        break;
      }
      case ('doAllPass' in result && !result.doAllPass): {
        this.setState({ chartValues: this.getChartValues(result) });
        resultMessageProps = {
          title: 'FAILURE',
          description: 'Stock does not pass financial ratio screens',
          isFailure: true,
        };
        break;
      }
      default:
        resultMessageProps = {
          title: 'ERROR',
          description: 'Something went wrong! Please try again later',
          isFailure: true,
        };
    }

    return resultMessageProps;
  }

  render() {
    const { resultMessageProps, chartValues, companyOptions } = this.state;

    return (
      <StyledHomePageDiv>
        <StyledImageHolder>
          <img src={TayyibStocksLogo} alt="Tayyib Stocks Logo" />
        </StyledImageHolder>
        <Fragment>
          <Dropdown
            search
            selection
            options={companyOptions}
            placeholder="Search for a stock..."
            onSearchChange={this.handleOnSearchChange}
            onChange={this.handleOnChange}
            onClick={this.handleOnClick}
          />
        </Fragment>
        {
          chartValues
          && (
            <ResultChart chartValues={chartValues} />
          )
        }
        {
          resultMessageProps
          && (
            <ResultMessage
              size="huge"
              {...resultMessageProps}
            />
          )
        }
      </StyledHomePageDiv>
    );
  }
}

export default HomePageContainer;
