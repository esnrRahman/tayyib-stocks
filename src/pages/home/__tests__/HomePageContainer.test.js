import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import HomePageContainer from '../HomePageContainer';
import {
  getCalculationFromYahoo, getCompanySymbolOptionsFromYahoo,
} from '../../../services/yahoo-finance-api-client';

jest.mock('../../../services/yahoo-finance-api-client');

describe('<HomePageContainer />', () => {
  let wrapper;
  let instance;

  beforeEach(() => {
    wrapper = shallow(<HomePageContainer />);
    instance = wrapper.instance();
  });

  it('intially renders correctly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders the ResultMessage if resultMessageProps is present', () => {
    const wrapperWithResult = shallow(<HomePageContainer />);

    wrapperWithResult.setState({
      resultMessageProps: {
        title: 'Test title',
        description: 'Test description',
        isFailure: false,
      },
    });

    expect(toJson(wrapperWithResult)).toMatchSnapshot();
  });

  it('renders the HomePageContainer if companyOptions are present', () => {
    const wrapperWithResult = shallow(<HomePageContainer />);

    wrapperWithResult.setState({
      companyOptions: [{
        key: '1',
        value: 'ABC',
        text: 'ABC company inc.',
      }],
    });

    expect(toJson(wrapperWithResult)).toMatchSnapshot();
  });

  it('handleOnChange', async () => {
    const testCompanyName = 'test company';
    const getCalculationFromYahooReturnValue = { b: 1 };
    const getResultMessageReturnValue = { a: 1 };
    getCalculationFromYahoo.mockImplementation(
      () => getCalculationFromYahooReturnValue
    );
    const preventDefaultFn = jest.fn();
    const getResultMessageFn = jest.fn(
      () => getResultMessageReturnValue
    );
    const setStateFn = jest.fn();
    const mockEvent = {
      preventDefault: preventDefaultFn,
    };
    const mockData = {
      value: testCompanyName,
    };


    instance.setState = setStateFn;
    instance.getResultMessage = getResultMessageFn;

    await instance.handleOnChange(mockEvent, mockData);

    expect(preventDefaultFn).toHaveBeenCalled();
    expect(setStateFn).toHaveBeenCalledWith(
      {
        yahooResult: null,
        resultMessageProps: null,
      }
    );
    expect(getCalculationFromYahoo).toHaveBeenCalledWith(testCompanyName);
    expect(getResultMessageFn).toHaveBeenCalledWith(
      getCalculationFromYahooReturnValue
    );
    expect(setStateFn).toHaveBeenCalledWith({
      resultMessageProps: getResultMessageReturnValue,
    });
  });

  it('handleOnSearchChange', async () => {
    const testQuery = 'test query';
    const getCompanySymbolOptionsFromYahooReturnValue = [
      {
        key: '1',
        value: 'ABC',
        text: 'ABC company',
      }];
    getCompanySymbolOptionsFromYahoo.mockImplementation(
      () => getCompanySymbolOptionsFromYahooReturnValue
    );
    const preventDefaultFn = jest.fn();
    const setStateFn = jest.fn();
    const mockEvent = {
      preventDefault: preventDefaultFn,
    };
    const mockData = {
      searchQuery: testQuery,
    };


    instance.setState = setStateFn;

    await instance.handleOnSearchChange(mockEvent, mockData);

    expect(preventDefaultFn).toHaveBeenCalled();
    expect(getCompanySymbolOptionsFromYahoo).toHaveBeenCalledWith(testQuery);
    expect(setStateFn).toHaveBeenCalledWith({
      companyOptions: getCompanySymbolOptionsFromYahooReturnValue,
    });
  });

  it('getResultMessage: Cannot find stock', () => {
    const cannotFindStockResult = {
      isFound: false,
    };

    const returnValue = instance.getResultMessage(cannotFindStockResult);

    expect(returnValue).toEqual(
      {
        title: 'ERROR',
        description: 'Unable to find the stocks for this company',
        isFailure: true,
      }
    );
  });

  it('getResultMessage: Stock Split', () => {
    const stockSplitResult = {
      doSplitsExist: true,
    };

    const returnValue = instance.getResultMessage(stockSplitResult);

    expect(returnValue).toEqual(
      {
        title: 'FAILURE',
        description: 'Cannot determine for this company as the stock has been split',
        isFailure: true,
      }
    );
  });

  it('getResultMessage: Successful calc', () => {
    const successResult = {
      doAllPass: true,
    };

    const returnValue = instance.getResultMessage(successResult);
    expect(returnValue).toEqual(
      {
        title: 'SUCCESS',
        description: 'The stock is considered halal',
        isFailure: false,
      }
    );
  });

  it('getResultMessage: Failed calc', () => {
    const inputResult = {
      doAllPass: false,
    };

    const returnValue = instance.getResultMessage(inputResult);
    expect(returnValue).toEqual(
      {
        title: 'FAILURE',
        description: 'The stock is not considered halal',
        isFailure: true,
      }
    );
  });

  it('getResultMessage: default return', () => {
    const inputResult = {
      somethingElse: 'defaultCase',
    };

    const returnValue = instance.getResultMessage(inputResult);
    expect(returnValue).toEqual(
      {
        title: 'ERROR',
        description: 'Something went wrong! Please try again later',
        isFailure: true,
      }
    );
  });
});
