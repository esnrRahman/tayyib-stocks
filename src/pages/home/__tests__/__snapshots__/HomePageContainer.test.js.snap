// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomePageContainer /> intially renders correctly 1`] = `
<styled.div>
  <styled.div>
    <img
      alt="Tayyib Stocks Logo"
      src="tstocks_full_logo.png"
    />
  </styled.div>
  <Input
    icon={
      <Icon
        as="i"
        circular={true}
        inverted={true}
        link={true}
        name="search"
        onClick={[Function]}
      />
    }
    onChange={[Function]}
    placeholder="Search for a stock..."
    size="massive"
    type="text"
  />
</styled.div>
`;

exports[`<HomePageContainer /> renders the ResultMessage if resultMessageProps is present 1`] = `
<styled.div>
  <styled.div>
    <img
      alt="Tayyib Stocks Logo"
      src="tstocks_full_logo.png"
    />
  </styled.div>
  <Input
    icon={
      <Icon
        as="i"
        circular={true}
        inverted={true}
        link={true}
        name="search"
        onClick={[Function]}
      />
    }
    onChange={[Function]}
    placeholder="Search for a stock..."
    size="massive"
    type="text"
  />
  <ResultMessage
    description="Test description"
    isFailure={false}
    size="huge"
    title="Test title"
  />
</styled.div>
`;
